avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,32 @@
 //! Serial Implementations
 
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
```

Comment:
All the other traits should also be implemented: `PartialEq`, `Eq`, `PartialOrd`, `Ord`, ...

Note that you'll have to implement this manually and calculate which one is bigger because `u2x` doubles the baudrate and this should be taken into account.

Reply:
Some of them implemented. I'm new to rust so please review and list other required.

Reply:
The right place to look for is the API guidelines:  It has a point [`C-COMMON-TRAITS`](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits) which lists all the important ones.

Looking at your implementation, I have a few notes:

1. I think you can remove the trait bound on `CLOCK` as you're not touching the type parameter anywhere.  It's only allowed to compare two `Baudrate`s if they have the same `CLOCK` anyway. So this is sufficient:

    ```rust
    impl<CLOCK> PartialEq for Baudrate<CLOCK> {
        fn eq(&self, other: &Self) -> bool {
            self.value() == other.value()
        }
    }
    ```

2. I think larger `ubrr` means smaller baudrate, but we need to recheck the datasheet on that ... If true, you got it the wrong way around (and also `u2x` should then mean _divide by 2_).

3. When `Ord` is implemented, you'd usually write `PartialOrd` like this:

    ```rust
    impl<CLOCK> PartialOrd for Baudrate<CLOCK> {
        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
            Some(self.cmp(other))
        }
    }
    ```

Reply:
1. Without bound I'm not able to use `value()` fn
1.
![32u4br](https://user-images.githubusercontent.com/1093324/96038314-3d6e6f00-0e67-11eb-8814-248199fdb301.png)
Looks like u2x decrease divider and increase baudrate.
1. Done
1. What abut asynchronous and synchronous difference? u2x works only on async. On sync always should be 0

Reply:
> Without bound I'm not able to use value() fn

You have to put `value()` into an impl block without the bound as well, then it'll work.  But honestly, it doesn't matter much ...

>  Looks like u2x decrease divider and increase baudrate.

Yes!  As per the formulas in the picture, _baudrate_ is inversely proportional to UBRR, so you need to swap operands in the `fn cmp()` impl (i.e. bigger UBRR then means smaller `Baudrate` value).  I think you'll also need to include the `+1` in your `fn value()` implementation to produce correct results.

> What abut asynchronous and synchronous difference? u2x works only on async. On sync always should be 0

We're always configuring the peripheral for _async_ operation:

https://github.com/Rahix/avr-hal/blob/7337cd76cd96f2d27701b137396d94a06d3a501d/avr-hal-generic/src/serial.rs#L71